### PR TITLE
Allow users to specify multiple adapters for workspaces

### PR DESCRIPTION
**Pull request by amstilp**
* Add an adapter registry where users can register custom adapters if desired.
* Add a workspace_type field to the `Workspace` model to indicate which adapter should be used.
* Modify urls to take the workspace_type as an argument.
* Add a mixin to a class variable containing the adapter for use in views.
* Modify templates to handle having multiple workspace types (and urls) instead of just one.
* Update docs.

### CODE REVIEW

**Comment by amstilp** [comment]
@jmcarson if you have a chance to review, here's a pull request implementing multiple adapters. If not, don't worry about it!

File: anvil_consortium_manager/apps.py [hunk 1]
```diff
@@ -1,9 +1,24 @@
...
+
+    def ready(self):
+        super().ready()
+        # Register specified adapters.
+        # Import here because importing outside of this method raises the AppRegistryNotReady exception.
+        from anvil_consortium_manager.adapters.workspace import (
+            workspace_adapter_registry,
+        )
+
+        adapter_modules = settings.ANVIL_WORKSPACE_ADAPTERS
```
Comment (jmcarson) [inline anvil_consortium_manager/apps.py:21]: Is it okay to have zero adapters? If not, you may want to validate the list is not empty.

**Review by jmcarson — APPROVED** [review]
Looks great. Just a few minor comments.

File: anvil_consortium_manager/views.py [hunk 2]
```diff
@@ -701,6 +730,7 @@ def post(self, request, *args, **kwargs):
         Handle POST requests: instantiate the forms instances with the passed
         POST variables and then check if they are valid.
         """
+        self.adapter = self.get_adapter()
```
[the remark this answers is not in the document]
Comment (amstilp) [reply anvil_consortium_manager/views.py:733]: For both WorkspaceCreate and WorkspaceImport, I had to change the post method from FormView enough that it didn't make sense to call super().post(), so I just set self.adapter here.

File: anvil_consortium_manager/apps.py (hunk 1 above)
Comment (amstilp) [reply anvil_consortium_manager/apps.py]: Good idea, I added a check.

File: anvil_consortium_manager/adapters/workspace.py [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+
+    def get_registered_names(self):
+        """Return a dictionary of registered adapter names."""
+        return {key: value().get_name() for (key, value) in self._registry.items()}
+
+
+# Initalize a global variable for the registry for use throughout the app.
+# Adapters will be added to the registry in the AppConfig for this app via a setting.
+workspace_adapter_registry = WorkspaceAdapterRegistry()
+"""Global variable to store all registsered workspace adapters.
```
[the remark this answers is not in the document]
Comment (amstilp) [reply anvil_consortium_manager/adapters/workspace.py]: fixed

File: anvil_consortium_manager/adapters/workspace.py [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+            else:
+                del self._registry[type]
+        else:
+            raise AdapterNotRegisteredError(
+                "adapter {} has not been registered yet.".format(adapter_class)
+            )
+
+    def get_adapter(self, type):
+        """ "Return an instance of the adapter for a given workspace ``type``."""
+        print(self._registry.keys())
```
[the remark this answers is not in the document]
Comment (amstilp) [reply anvil_consortium_manager/adapters/workspace.py]: fixed

**Pull request merged**